### ISSUE TITLE
How to correspond GT and and labels for eval?

### ISSUE THREAD

**Issue by ilikeokoge** [yukimasano/self-label#13]
Thanks for your codebase!

I tried cifar.py  for cifar 10 and did not know bellow code.

acc = kNN(model, trainloader, testloader, K=10, sigma=0.1, dim=knn_dim)

kNN devide test data of cifar10 to 10 clusters and calclate acc with GT?
If so, how to correspond GT classes and and clusters for eval?

**Comment by yukimasano** [comment]
Hi there!

as noted somewhere, the kNN code is from this repo: https://github.com/Raymond-sci/AND, (more specifically, here: https://github.com/Raymond-sci/AND/blob/master/lib/protocols.py). 

In the case of kNN ("K" nearest neighbors), the "K" refers to how many neighbors you use for interpolating the class.
As noted in the Paper, for CIFAR-10 we use 128 clusters (for pretraining -- not for evaluation) though, as we always use overclustering.

**Comment by ilikeokoge** [comment]
Thanks for your reply!
Yes, I know you use 128 clusters for pretraining. But how about eval?
`if args.test_only:
    feature_return_switch(model, True)
    usepca = True
    acc = kNN(model, trainloader, testloader, K=[200, 50, 10, 5, 1], sigma=[0.1, 0.5], dim=knn_dim, use_pca=usepca)
    sys.exit(0)`

When you use kNN for eval, do you use GT of test dataset(10class)?
Or you do not GT for eval (and off course trainng) and you use pseudo label gained in pretraing?

**Comment by ilikeokoge** [comment]
Thanks for your reply!
Yes, I know you use 128 clusters for pretraining. But how about eval?
`if args.test_only: feature_return_switch(model, True) usepca = True acc = kNN(model, trainloader, testloader, K=[200, 50, 10, 5, 1], sigma=[0.1, 0.5], dim=knn_dim, use_pca=usepca) sys.exit(0)`

When you use kNN for eval, do you use GT of test dataset(10class)?
Or you do not GT for eval (and off course trainng) and you use pseudo label gained in pretraing?

**Comment by yukimasano** [comment]
Yes, for kNN you use the labels of the training set and (say if k=1) you assign your test-sample the label of the nearest neighbor. You then check if this matches the label of the ground-truth, and that's how kNN accuracy is calculated.

In this paper we mostly focussed on representation learning, so we mostly only analyze how good the features are, not how good the pseudo-labels are (although I recently ran some baselines, and they are quite good 😄 .
